docs/spark.rst

```diff
@@ -0,0 +1,106 @@
+mounted file system, that ensures that a given file can be accessed by all of
+the workers in the cluster. We run Spark on top of HDFS, which is a typical
+configuration. However, Spark can also load data from block stores like S3, or
+from HPC style distributed file systems like NFS, Gluster, or Ceph. In HDFS,
+the master is called the namenode, and workers are datanodes. The HDFS namenode
+manages file metadata across the cluster, while the datanodes store the raw data.
+HDFS has very good performance for large reads; write performance varies with
+the cluster configuration.
+
+We decompose this cluster into two separate classes inside of Toil:
+:class:`toil.lib.spark.SparkService` and
+:class:`toil.lib.spark.WorkerService`. The
+:class:`toil.lib.spark.SparkService` class launches the Spark master
+and HDFS namenode, and then adds one or more
+:class:`toil.lib.spark.WorkerService` child services. The
+:class:`toil.lib.spark.WorkerService` class then launches the
+Spark worker and HDFS datanode. The individual Spark and HDFS applications are
+run using docker containers from cgl-docker-lib_.
+
+.. _cgl-docker-lib: https://github.com/BD2KGenomics/cgl-docker-lib
+
+Launching a Spark Cluster
+-------------------------
+
+To launch a Spark cluster, import:
+
+::
+        from toil_lib.spark import spawn_spark_cluster
+
+Then, add a service job:
+
+::
+        masterIPPromise = spawn_spark_cluster(job, ...)
+
+The service job will return a promise for the IP address of the master node.
+The service job takes several parameters:
+
+- job: The job that the Spark cluster should descend from.
+- sudo: Whether or not to run the Docker containers with sudo permission.
+- numWorkers: The number of workers to start in the cluster. Must be greater than
+  or equal to one.
+
+We also take several optional parameters:
+
+- cores: An optional parameter to set the number of cores that each machine should
+  use. If not provided, we will auto-detect the cores on a worker.
+- overrideLeaderIP: an optional parameter that can override the leader IP.
+- memory: The amount of memory to allocate for each worker used by Spark.
+- disk: The amount of disk to request for each worker.
+
+When launching a Spark cluster, there is one important caveat: the job that
+launches the services must be executed with `checkpoint = True`. If the job is
+not checkpointable, then the cluster may not restart robustly. This is important
+if you are running on a preemptable cluster: in Spark and HDFS, failed workers can
+be restarted, but a failed master will cause all workers in the cluster to
+dissociate. If the cluster does not restart, this causes a deadlock.
+
+Using Locally
+-------------
+
+To use Spark locally, you must install Toil, toil-lib, and Docker. Docker installation
+instructions can be found [here](https://docs.docker.com/engine/installation/).
+For examples of how to connect to a Toil-based Spark cluster, see the Spark
+unit tests:
+
```

Comment:
I'm in favor. I'm familiar with about half of the functions in `toil_lib/programs.py`, so I could get it started but would likely need some help.